### DISCUSSION TITLE
Error validation for usercontrol properties

### DISCUSSION THREAD

**Discussion by trashhacker** [AvaloniaUI/Avalonia#17130] [Q&A] [answered]
How to make User Control Show validations Like textbox

**Comment by stevemonaco** [comment]
Check the `TextBox` control for implementation details. ie. `DataValidationErrors` control: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/TextBox.xaml#L110-L124

However, I typically just expose errors as properties on the VM and have a `UserControl` show it through a `TextBlock` though. It doesn't make sense to use the same approach as lookless controls to me.

**Reply by trashhacker** [reply]
thanks

**Comment by trashhacker** [comment]
![Screenshot 2024-09-26 101837](https://github.com/user-attachments/assets/ba2ed699-a3ff-4f40-bfe0-da3010df621a)

i want this what the textbox has for all usercontrols, how to archive that 
could u post me a link to the default usercontrol controltemplate?
and tell me how to override the default style controltemplate globaly

**Comment by trashhacker** [comment, marked as answer]
Oh lord figured it out xD why do u have not implemented it as default why i have to implement it myself ? ...


    public static readonly StyledProperty<string> UserTextProperty = 
        AvaloniaProperty.Register<TestUserControl, string>(nameof(UserText), defaultBindingMode:BindingMode.TwoWay, enableDataValidation:true);
    public string UserText
    {
        get => GetValue(UserTextProperty);
        set => SetValue(UserTextProperty, value);
    }
    public TestUserControl()
    {
        InitializeComponent();
    }
    protected override void UpdateDataValidation(AvaloniaProperty property, BindingValueType state, Exception? error)
    {
        if (property == UserTextProperty)
        {
            DataValidationErrors.SetError(this, error);
        }
    }

    <ControlTheme x:Key="{x:Type controls:TestUserControl}" TargetType="{x:Type controls:TestUserControl}">
      <Setter Property="BorderBrush" Value="Yellow"/>
      <Setter Property="BorderThickness" Value="2"/>
      <Setter Property="Template">
        <ControlTemplate>
          <DataValidationErrors >
            <Border BorderThickness="{TemplateBinding BorderThickness}" BorderBrush="{TemplateBinding BorderBrush}">
              <ContentPresenter Name="PART_ContentPresenter" Content="{TemplateBinding Content}"/>
            </Border>
          </DataValidationErrors >
        </ControlTemplate>
      </Setter>
    </ControlTheme>

**Reply by stevemonaco** [reply]
Most controls do not participate in error validation, so it's an unnecessary cost to force even more required elements into the visual tree that will need to be measured, rendered, or handle events. As noted by your code, you glued together a few small components that did the work for you which is far less code than a from-scratch implementation. You're free to encapsulate this into your own derived `UserControl` type for reusability.

**Reply by trashhacker** [reply]
im sorry (: i was frustrated ...
but it would be very helpfull to write a better documentation on this topic.

**Reply by timunie** [reply]
> but it would be very helpfull to write a better documentation on this topic.

@trashhacker feel free to open a PR on the docs. We need community support to iron out the docs.

**Reply by trashhacker** [reply]
how can i do that will a word document help u out ? or do u prefere a other format
i tried to find out how to make the pull request but the other request are not helfpfull :/

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/avalonia-docs/tree/main/docs here you have all the md files where you can change the content.
